src/gtpu/encap.c

```diff
@@ -345,15 +346,84 @@ static int gtp5g_buf_skb_encap(struct sk_buff *skb, struct net_device *dev,
             return -1;
         }
 
-        if (unix_sock_send(pdr, skb->data, skb_headlen(skb), 0) < 0) {
-            GTP5G_ERR(dev, "Failed to send skb to unix domain socket PDR(%u)", pdr->id);
-            ++pdr->ul_drop_cnt;
+        if (pdr_addr_is_netlink(pdr)) {
+            if (netlink_send(pdr, skb, dev_net(dev)) < 0) {
+                GTP5G_ERR(dev, "Failed to send skb to netlink socket PDR(%u)", pdr->id);
+                ++pdr->dl_drop_cnt;
+            }
+        } else {
+            if (unix_sock_send(pdr, skb->data, skb_headlen(skb), 0) < 0) {
+                GTP5G_ERR(dev, "Failed to send skb to unix domain socket PDR(%u)", pdr->id);
+                ++pdr->ul_drop_cnt;
+            }
         }
     }
     dev_kfree_skb(skb);
     return 0;
 }
 
+/* Function netlink_{...} are used to handle buffering */
+// Send PDR ID, FAR action and buffered packet to user space
+static int netlink_send(struct pdr *pdr, struct sk_buff *skb_in, struct net *net)
+{
+    struct sk_buff *skb;
+    static atomic_t seq_counter;
+    u32 seq;
+    void *header;
+    struct nlattr *attr;
+    int err;
+
+    skb = genlmsg_new(
+        nla_total_size_64bit(8) +
+            nla_total_size(2) +
+            nla_total_size(2) +
+            nla_total_size(skb_in->len),
+        GFP_ATOMIC);
+    if (!skb)
+        return -ENOMEM;
+
+    seq = atomic_inc_return(&seq_counter);
+    header = genlmsg_put(skb, 0, seq, &gtp5g_genl_family, 0, GTP5G_CMD_BUFFER_GTPU);
+    if (!header)
+    {
+        nlmsg_free(skb);
+        return -ENOMEM;
+    }
+
+    err = nla_put_u16(skb, GTP5G_BUFFER_ID, pdr->id);
```

Comment:
@ShouheiNishi 

The latest report format is changed.
Please refer to the latest `unix_sock_send()` to modify nlmsg format to support DLDR & USAR report.
Thanks.

Reply:
Since the USAR reports and the DLDR reports can be distinguished by the cmd number in nlmsg, we believe that there is no need to change the msg format.